### PR TITLE
Teach Aggregable[T] how to groupBy

### PR DESCRIPTION
**Pull request by danking**
AST now can be asked to produce a pair of a TypedAggregator and a CMCPSCode[T]. The former is the Aggregator corresponding to the terminal expression, the latter represents the series of transformations to the aggregable that should be applied before the aggregator's sequence op.

The syntax looks like:

```
gs.groupBy(x => sa.population, gs => gs.hardyWeinberg())
```

Nested `groupBy` is permitted:

```
gs.groupBy(x  => sa.population,
           gs => gs.groupBy(x  => sa.subPopulation,
                            gs => gs.hardyWeinberg())
```

Addresses #878

### CODE REVIEW

File: src/main/scala/is/hail/expr/FunctionRegistry.scala [hunk 1]
```diff
@@ -237,10 +226,109 @@ object FunctionRegistry {
...
+          key = (x: Any) => {
+            localA(idx) = x
+            keyBodyFn(localA.asInstanceOf[mutable.ArrayBuffer[AnyRef]])
+          };
+
+          (downstreamParamId, localA2) <- ecNewPosition();
+
+          downstreamParamTypeFixed = {
+            val x = downstreamParamType.asInstanceOf[TAggregable].copy();
+            // FIXME filter out old bindings
```
Comment (danking) [inline src/main/scala/is/hail/expr/FunctionRegistry.scala]: @tpoterba You commented on the other PR:
> does this work?
I think it does, but for terrible reasons. I'm really beginning to think Aggregators need to be rethought because I can barely keep my head around them.

So how bindings normally work is that you have your normal EvalContext which for variant parallel stuff contains things like `global`, `v`, `va`, and `gs`. The `gs` binding is special: it has an internal symbol table that *must* (although this isn't completely enforced) overlap exactly with the outer table, but may contain extra bindings. For variant aggregations, this means it has `global`, `v`, `va`, and `g` (which occupies the same position as `gs`), but it additionally has `s` and `sa`.

So, this looks more or less ok, but I think I may have fucked things up and/or made it way more complicated when I changed aggregators to be in "continuation passing style". I intended to simplify things, but I think I made it worse. For example, what happens when folks use a lambda with their own custom name for `g`?

```scala
vds.annotateVariantsExpr('gs.map(cake => cake.gt).count()')
```

In this case we have to add a binding for `cake` that points to wherever `g` is. I'm not sure what we were doing before my change. After my change, every transformation (i.e. `map`) takes the value of `g` (to give to the lambda) and takes a lambda `Any => Unit` which is the "continuation". It's whatever should happen after the `map` function runs. In this case, that means it just calls `seqOp` on the count aggregator.

Here, this "take the value of `g`" and give it to lambda is done by `keyFn`. It set's a position in the eval context's array buffer to its argument. The `keyBodyFn` knows that `keyParam` (which in this example is `cake`) is at position `idx`.

So this seems complex and not exactly right, but it works.

Now, I think with `groupBy`'s nested aggregators things get even more complex and maybe not even right any more. Here we basically want the exact same *outer* symbol table, but now `gs` should be some new binding. For example:

```scala
vds.annotateVariantsExpr('gs.groupBy(g => sa.population, newGs => newGs.count())')
```

here we want the bindings of the second lambda to still include `global`, `va`, and `v`, but not include `gs` and do include `newGs`.

I think you're right that I've got it wrong, but I need to think through this more to get it right. This is really pushing me towards a clean slate implementation of expr, because I feel like the complexity is getting ahead of me now.

Comment (tpoterba) [inline src/main/scala/is/hail/expr/FunctionRegistry.scala]: Very good points. I'm still happy to put this in in the meantime, though, but would like to understand a little better to anticipate edge cases.

Comment (tpoterba) [inline src/main/scala/is/hail/expr/FunctionRegistry.scala]: How are you feeling about this today? Can you add the `cake` test into a suite, preferably with a few levels of nesting?

**Comment by danking** [comment]
@cseed @tpoterba I added several tests around binding that make me feel much more comfortable with the correctness.

**Comment by danking** [comment]
@cseed @tpoterba I added several tests around binding that make me feel much more comfortable with the correctness.

File: src/main/scala/is/hail/expr/FunctionRegistry.scala [hunk 2]
```diff
@@ -238,10 +227,108 @@ object FunctionRegistry {
...
+            fatal(s"Argument evaluated to missing in call to aggregator $name"))
+
+        ) yield aggregator.ctor(g, v.asInstanceOf[v])
+
+      case aggregator: GroupByAggregatorFun[t] =>
+        val Lambda(_, keyParam, keyBody) = args(1)
+        val TFunction(Seq(keyParamType), _) = argTypes(1)
+
+        val Lambda(_, downstreamParam, downstreamBody) = args(2)
+        val TFunction(Seq(downstreamParamType), _) = argTypes(2)
```
Comment (cseed) [inline src/main/scala/is/hail/expr/FunctionRegistry.scala]: You can put a type constraint `downstreamParamType: TAggregable` and remove the casts below.

File: src/main/scala/is/hail/expr/FunctionRegistry.scala [hunk 3]
```diff
@@ -238,10 +227,108 @@ object FunctionRegistry {
...
+
+          (downstreamParamId, localA2) <- ecNewPosition();
+
+          downstreamParamTypeFixed = {
+            val x = downstreamParamType.asInstanceOf[TAggregable].copy();
+            x.symTab = downstreamParamType.asInstanceOf[TAggregable].symTab + ((downstreamParam, (downstreamParamId, downstreamParamType.asInstanceOf[TAggregable].elementType)))
+            x
+          };
+
+          (continuationId, localA3) <- ecNewPosition();
```
Comment (cseed) [inline src/main/scala/is/hail/expr/FunctionRegistry.scala]: Idx instead of Id?  I was confusing the `a` index for the param name below.

File: src/main/scala/is/hail/expr/AST.scala [hunk 4]
```diff
@@ -574,6 +588,15 @@ case class ApplyMethod(posn: Position, lhs: AST, method: String, args: Array[AST
     lhs.typecheck(ec)
     (lhs.`type`, method, args) match {
 
+      case (it: TAggregable, "groupBy", Array(Lambda(_, param1, body1), Lambda(_, param2, body2), rest@_*)) =>
+        rest.foreach(_.typecheck(ec.copy(st = emptySymTab)))
+        body1.typecheck(ec.copy(st = it.symTab + ((param1, (-1, it.elementType)))))
+        body2.typecheck(ec.copy(st = ec.st.filter { case (name, (idx, typ)) => !typ.isInstanceOf[TAggregable] } + ((param2, (it.symTab.find { case (_, (_, t)) => t == it.elementType }.get._2._1, it)))))
```
Comment (cseed) [inline src/main/scala/is/hail/expr/AST.scala]: As per our discussion, we need to allow for multiple aggregations inside the second expression in the groupBy.

**Comment by danking** [comment]
Closing until I have time to rewrite using recursively passed EvalContexts.
